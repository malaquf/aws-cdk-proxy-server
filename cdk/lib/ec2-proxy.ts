import {Fn, Stack, StackProps} from "aws-cdk-lib";
import {Construct} from "constructs";
import {
    AmazonLinuxImage,
    CfnInstance,
    InstanceClass,
    InstanceSize,
    InstanceType,
    Peer,
    Port,
    SecurityGroup,
    UserData,
    Vpc
} from "aws-cdk-lib/aws-ec2";
import {CfnInstanceProfile, ManagedPolicy, Role, ServicePrincipal} from "aws-cdk-lib/aws-iam";
import { StringParameter } from "aws-cdk-lib/aws-ssm";

export class Ec2Stack extends Stack {

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        /**
         * THIS STACK CURRENTLY CAN'T BE USED AT THE MOMENT
         * BECAUSE CDK IS NOT ATTACHING THE
         * INSTANCE PROFILE TO THE INSTANCE
         */

        const vpc = Vpc.fromLookup(this, 'ImportedVpc', {
            vpcId: StringParameter.valueFromLookup(this, '/VpcId'),
        });

        const ec2SecurityGroup = new SecurityGroup(this, "Ec2SecurityGroup", {
            vpc: vpc,
            description: "Security Group for Ec2 Proxy instance",
            disableInlineRules: true,
            securityGroupName: "Ec2ProxySecurityGroup",
            allowAllOutbound: false,
            allowAllIpv6Outbound: false,
        });
        ec2SecurityGroup.addEgressRule(Peer.anyIpv4(), Port.tcp(3128), 'squid proxy');

        const instanceRole = new Role(this, 'EC2ProxyRole', {
            roleName: 'EC2ProxyRole',
            description: 'EC2 Proxy Role',
            managedPolicies: [ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore')],
            assumedBy: new ServicePrincipal('ec2.amazonaws.com')
        });

        const profile = new CfnInstanceProfile(this, `InstanceProfile`, {
            roles: [ instanceRole.roleName ]
        });

        const ssmaUserData = UserData.forLinux();
        ssmaUserData.addCommands(
            'yum update -y -q',
            'yum install ec2-instance-connect',
            'sudo systemctl enable amazon-ssm-agent',
            'sudo systemctl start amazon-ssm-agent',
            'sudo yum install squid',
            'sudo service squid restart'
        );

        new CfnInstance(this, 'ProxyServerInstance', {
            networkInterfaces: [
                {
                    deviceIndex: "0",
                    subnetId: vpc.privateSubnets[0].subnetId
                }
            ],
            iamInstanceProfile: profile.ref,
            imageId: new AmazonLinuxImage().getImage(this).imageId,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString(),
            userData: Fn.base64(ssmaUserData.render()),
            tags: [{key: 'Name', value: 'proxy-server'}],
            });
    }

}