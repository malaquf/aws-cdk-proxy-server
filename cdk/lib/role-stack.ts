import { Construct } from 'constructs';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import {Stack, StackProps, Tags} from 'aws-cdk-lib';
import { InterfaceVpcEndpointAwsService, SubnetType, Vpc } from 'aws-cdk-lib/aws-ec2';
import { AccountPrincipal, Effect, PolicyDocument, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';

export class RoleStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);
        
        const role = new Role(this, 'ProxyUserRole', {
            assumedBy: new AccountPrincipal(this.account),
            roleName: 'proxy-user-role',
            description: 'Role for accessing proxy server',
        });
        role.addToPolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
                'ec2:DescribeInstances'
            ],
            resources: ['*']
        }));
        role.addToPolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
                'ssm:StartSession',
                'ec2-instance-connect:SendSSHPublicKey'
            ],
            resources: ['arn:aws:ec2:*:*:instance/*'],
            conditions: [
                { 'StringEquals' : { 'aws:ResourceTag/Name': 'proxy-server' }}
            ]
        }));
        role.addToPolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
                'ssm:StartSession'
            ],
            resources: ['arn:aws:ssm:*:*:document/AWS-StartSSHSession']
        }));
        role.addToPolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
                'ssm:TerminateSession',
                'ssm:ResumeSession'
            ],
            resources: ['arn:aws:ssm:*:*:session/$${aws:username}-*']
        }));
    }
}
