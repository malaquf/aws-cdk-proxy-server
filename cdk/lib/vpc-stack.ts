import { Construct } from 'constructs';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import {Stack, StackProps, Tags} from 'aws-cdk-lib';
import { InterfaceVpcEndpointAwsService, SubnetType, Vpc } from 'aws-cdk-lib/aws-ec2';

export class VpcStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);
        
        const vpc = new Vpc(this, 'Vpc', {
            maxAzs: 1, // increase it for resilience
            subnetConfiguration: [
                {
                    cidrMask: 24,
                    name: 'public',
                    subnetType: SubnetType.PUBLIC,
                },
                {
                    cidrMask: 24,
                    name: 'private',
                    subnetType: SubnetType.PRIVATE_WITH_EGRESS,
                }
            ]
        });

        // allows access to EC2 in private subnet through SSM
        vpc.addInterfaceEndpoint('SSMInterfaceVpcEndpoint', {
            service: new InterfaceVpcEndpointAwsService('ssm'),
            subnets: { subnetType: SubnetType.PUBLIC},
            privateDnsEnabled: true,
            open: true
        });
        vpc.addInterfaceEndpoint('Ec2InterfaceVpcEndpoint', {
            service: new InterfaceVpcEndpointAwsService('ec2'),
            subnets: { subnetType: SubnetType.PUBLIC},
            privateDnsEnabled: true,
            open: true
        });
        vpc.addInterfaceEndpoint('SSMMessagesInterfaceVpcEndpoint', {
            service: new InterfaceVpcEndpointAwsService('ssmmessages'),
            subnets: { subnetType: SubnetType.PUBLIC},
            privateDnsEnabled: true,
            open: true
        });
        vpc.addInterfaceEndpoint('EC2MessagesInterfaceVpcEndpoint', {
            service: new InterfaceVpcEndpointAwsService('ec2messages'),
            subnets: { subnetType: SubnetType.PUBLIC},
            privateDnsEnabled: true,
            open: true
        });

        new StringParameter(this, 'VpcId', {
            parameterName: '/VpcId',
            stringValue: vpc.vpcId
        });

    }
}
