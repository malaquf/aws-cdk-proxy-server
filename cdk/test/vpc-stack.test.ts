import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { VpcStack } from '../lib/vpc-stack';


describe('vpc stack - snapshot test', () => {

    it('stack is still valid', () => {
        const app = new App();

        const stack = new VpcStack(app, 'MyTestStack');

        expect(Template.fromStack(stack)).toMatchSnapshot();
    });
});
