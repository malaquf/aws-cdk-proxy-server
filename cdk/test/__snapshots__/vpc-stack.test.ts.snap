// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vpc stack - snapshot test stack is still valid 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Vpc8378EB38": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcEC2MessagesInterfaceVpcEndpointC9727455": {
      "Properties": {
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "VpcEC2MessagesInterfaceVpcEndpointSecurityGroup32EC6140",
              "GroupId",
            ],
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".ec2messages",
            ],
          ],
        },
        "SubnetIds": [
          {
            "Ref": "VpcpublicSubnet1Subnet2BB74ED7",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEC2MessagesInterfaceVpcEndpointSecurityGroup32EC6140": {
      "Properties": {
        "GroupDescription": "MyTestStack/Vpc/EC2MessagesInterfaceVpcEndpoint/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::GetAtt": [
                "Vpc8378EB38",
                "CidrBlock",
              ],
            },
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "Vpc8378EB38",
                      "CidrBlock",
                    ],
                  },
                  ":443",
                ],
              ],
            },
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc",
          },
        ],
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VpcEc2InterfaceVpcEndpoint1F8C0BDB": {
      "Properties": {
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "VpcEc2InterfaceVpcEndpointSecurityGroup2A01BA68",
              "GroupId",
            ],
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".ec2",
            ],
          ],
        },
        "SubnetIds": [
          {
            "Ref": "VpcpublicSubnet1Subnet2BB74ED7",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEc2InterfaceVpcEndpointSecurityGroup2A01BA68": {
      "Properties": {
        "GroupDescription": "MyTestStack/Vpc/Ec2InterfaceVpcEndpoint/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::GetAtt": [
                "Vpc8378EB38",
                "CidrBlock",
              ],
            },
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "Vpc8378EB38",
                      "CidrBlock",
                    ],
                  },
                  ":443",
                ],
              ],
            },
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc",
          },
        ],
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VpcIGWD7BA715C": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VpcId84E18EF0": {
      "Properties": {
        "Name": "/VpcId",
        "Type": "String",
        "Value": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "VpcSSMInterfaceVpcEndpoint4879EC87": {
      "Properties": {
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "VpcSSMInterfaceVpcEndpointSecurityGroupEA29E658",
              "GroupId",
            ],
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".ssm",
            ],
          ],
        },
        "SubnetIds": [
          {
            "Ref": "VpcpublicSubnet1Subnet2BB74ED7",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcSSMInterfaceVpcEndpointSecurityGroupEA29E658": {
      "Properties": {
        "GroupDescription": "MyTestStack/Vpc/SSMInterfaceVpcEndpoint/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::GetAtt": [
                "Vpc8378EB38",
                "CidrBlock",
              ],
            },
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "Vpc8378EB38",
                      "CidrBlock",
                    ],
                  },
                  ":443",
                ],
              ],
            },
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc",
          },
        ],
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VpcSSMMessagesInterfaceVpcEndpointB81AE340": {
      "Properties": {
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "VpcSSMMessagesInterfaceVpcEndpointSecurityGroupE0AAAFE7",
              "GroupId",
            ],
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".ssmmessages",
            ],
          ],
        },
        "SubnetIds": [
          {
            "Ref": "VpcpublicSubnet1Subnet2BB74ED7",
          },
        ],
        "VpcEndpointType": "Interface",
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcSSMMessagesInterfaceVpcEndpointSecurityGroupE0AAAFE7": {
      "Properties": {
        "GroupDescription": "MyTestStack/Vpc/SSMMessagesInterfaceVpcEndpoint/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::GetAtt": [
                "Vpc8378EB38",
                "CidrBlock",
              ],
            },
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "Vpc8378EB38",
                      "CidrBlock",
                    ],
                  },
                  ":443",
                ],
              ],
            },
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc",
          },
        ],
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VpcVPCGWBF912B6E": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "VpcIGWD7BA715C",
        },
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "VpcprivateSubnet1DefaultRouteB506891A": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "VpcpublicSubnet1NATGatewayA036E8A6",
        },
        "RouteTableId": {
          "Ref": "VpcprivateSubnet1RouteTable1979EACB",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcprivateSubnet1RouteTable1979EACB": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/privateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcprivateSubnet1RouteTableAssociationEEBD93CE": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VpcprivateSubnet1RouteTable1979EACB",
        },
        "SubnetId": {
          "Ref": "VpcprivateSubnet1SubnetCEAD3716",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcprivateSubnet1SubnetCEAD3716": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/privateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcpublicSubnet1DefaultRouteB88F9E93": {
      "DependsOn": [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": {
          "Ref": "VpcpublicSubnet1RouteTable15C15F8E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcpublicSubnet1EIP411541E6": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/publicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VpcpublicSubnet1NATGatewayA036E8A6": {
      "DependsOn": [
        "VpcpublicSubnet1DefaultRouteB88F9E93",
        "VpcpublicSubnet1RouteTableAssociation4E83B6E4",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "VpcpublicSubnet1EIP411541E6",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "VpcpublicSubnet1Subnet2BB74ED7",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/publicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VpcpublicSubnet1RouteTable15C15F8E": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/publicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcpublicSubnet1RouteTableAssociation4E83B6E4": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VpcpublicSubnet1RouteTable15C15F8E",
        },
        "SubnetId": {
          "Ref": "VpcpublicSubnet1Subnet2BB74ED7",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcpublicSubnet1Subnet2BB74ED7": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "MyTestStack/Vpc/publicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
